### ISSUE TITLE
24.0 upgrade issue when LDAP edit-mode set to empty

### ISSUE THREAD

**Issue by mqu** [keycloak/keycloak#28446]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

ldap

### Describe the bug

upgrading from 23.0.7 to 24.0.2 is not possible, when some LDAP user-federation setting have "edit-mode" set to empty.

### Version

24.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

when trying to upgrade from 23.0.7 to 24.0.2, I would like to have upgrade working smoothly.

### Actual behavior

With some REALM, connected to LDAP user federation, with edit-mode set as blank, when I run upgrade process, I have an exception triggered : ldapErrorEditModeMandatory.
[keycloak-error-24.0.2.log](https://github.com/keycloak/keycloak/files/14873609/keycloak-error-24.0.2.log)

### How to Reproduce?

before upgrading to 24.0, setup LDAP user federation with "edit-mode" to blank. Then, try to upgrade to 24.0 version.

### Anything else?

setting edit-mode to "read-only" in each realm and each "user-federation" with LDAP will ensure you can upgrade safely. Empty setting should not be possible ; I don't know why I have been facing this issue.

best regards, Marc.

**Comment by pedroigor** [comment]
Were you able to migrate from other versions when the edit mode is not set?

**Comment by mqu** [comment]
yes, in did !

**Comment by arattur** [comment]
I have run into same problem in my **TEST** environment. I'm not able to start Keycloak after upgrade 23.0.6 => 24.0.2 with the same error. Only that in my case I had **before** and have **after** **migration** in every realm where user federation through ldap is enabled following values in _component_config_: 
`editMode  = ’READ_ONLY’ AND validatePasswordPolicy = ’false’ AND useTruststoreSpi = ’ldapsOnly’ `

I added _--verbose_ to _kc.sh_ and receive following error: org.keycloak.storage.ldap.LDAPStorageProviderFactory.validateConfiguration(LDAPStorageProviderFactory.java:293)
[error_log.txt](https://github.com/keycloak/keycloak/files/14882469/error_log.txt)

Before upgrading my **TEST** environment I somehow (actually I hade some initial errors due to configuration changes between v23 and v24) managed to upgrade successfully my **DEV** environment. Same Keycloak versions. I have tried to find out what are differences and I have spotted that:
- In DEV after successfull upgrade the value of `useTruststoreSpi` in _component_config_ table is set to `always`, which is expected outcome according to: https://github.com/keycloak/keycloak/blob/24.0.2/model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo24_0_0.java#L100
- In TEST I can see, that Liquibase migration is done. The value of `useTruststoreSpi` is still `ldapsOnly`, ie value change to `allways` has not been done by MigrateTo24_0_0.java

Using Keycloak docker [image](https://hub.docker.com/layers/keycloak/keycloak/24.0.2-0/images/sha256-aefd0843ff5f4800df05c6fc48af3f8aa50e319451fa4a0f2417ba2b938ed7de?context=explore) , running on Kubernetes.

**Comment by sguilhen** [comment]
@arattur Can you share your LDAP provider config in 23.0.6? If I got it right, you had a valid config (`editMode` set to `READ_ONLY`) that somehow got invalid after the upgrade? The check for a valid edit mode has been in the codebase for years now, it shouldn't be possible to have a config in 23 with that value set to empty.

**Comment by sguilhen** [comment]
If you had an empty `editMode` in one of the LDAP providers (which you shouldn't have), then the [migration to 24](https://github.com/keycloak/keycloak/blob/24.0.2/model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo24_0_0.java) will cause the error for you because it updates the LDAP storage provider component, and the validation is performed every time the provider is updated. So an invalid state will be flagged and the migration will fail.

To fix this, you have to update the LDAP config in KC 23 to use one of the available edit modes.

**Comment by mqu** [comment]
when I grep editMode in SQL backup, I have never seen editMode = null, but allways something like this:
```
c37317b3-cf1d-4dd4-97cd-ef6f5c393593	ea2396aa-9906-4071-9e5d-ca9bbb704df3	editMode	READ_ONLY
...
6ec21ce6-73e4-4576-95b9-383dc0ab08ce	1308c6d5-aa81-46b4-92d7-05077bad5b3e	editMode	UNSYNCED
8e759370-5fd5-4586-b911-454dc5aa1e1d	fe8b65da-8d3b-4343-85eb-94196a331421	editMode	READ_ONLY
```
- our backup history go back to nearly one year,
- we have upgraded sequentially from : 20.0.2, 20.0.5, 22.01, 22.0.5, 23.0.1, 23.0.7, never having issues with editMode before.

**Comment by pedroigor** [comment]
Can we close this one then?

**Comment by sguilhen** [comment]
@mqu I'm just a bit confused with the issue description, where you mention in steps to reproduce "_before upgrading to 24.0, setup LDAP user federation with "edit-mode" to blank. Then, try to upgrade to 24.0 version._", which led me to believe you have at least one provider with an empty value for the `editMode` before you migrated.

All other migrations (from 20.x to 21.x to 22.x, etc) don't fail because the component model of the LDAP providers is not updated, so they don't trigger the edit mode validation. In 24, we updated the value of the `useTruststoreSpi` attribute in the providers because the `ldapsOnly` value doesn't exist anymore. This triggers a component update that in turn triggers the validation that will flag the inconsistent state of the LDAP provider in case the edit mode was empty before the migration.

**Comment by mqu** [comment]
when I have been seeking for `editMode` for LDAP connections, I found two of them unset, and all the other where on `READ_ONLY` as expected. For those 2 unset one, after setting them to READ_ONLY, and save database, I have been able to upgrade smootly on 24.0 version.

**Comment by sguilhen** [comment]
@mqu Good to know. What I believe happened was that some earlier version of Keycloak allowed empty edit mode, so if you started back then and migrated with every release you could have a provider with this value unset at this point, which causes the migration to 24 to fail.

If that's the case, then we might have missed a migration step back when we introduced the requirement edit mode can't be empty to migrate old providers to READ_ONLY where necessary. I know it is inconvenient, but the workaround in this case is to do exactly what you did - locate the providers with the empty value and set it before migrating.

**Comment by mqu** [comment]
@sguilhen : why not test editMode and set it to READ_ONLY when it's value is NULL or empty ?

now, having `ldapErrorEditModeMandatory` exception, it's not too difficult to gather information after searching on the Internet. Just seek at each LDAP setting for edit-mode having "null" or "empty" value.

**Comment by sguilhen** [comment]
@mqu Yeah, we can add a migration for that, and will probably do so - but it won't help those who already migrated and hit the issue like yourself. I'll discuss this later today with the team to decide on the best approach to handle this.

**Comment by pedroigor** [comment]
@mqu @arattur Thanks for the additional info as they help to sort out what might have happened.

For now, I'm closing this issue but we'll keep an eye on similar issues to see if we should actually fix our migration logic to avoid such errors.
